staging/kos/pkg/controllers/connectionagent/connection_agent.go

```diff
@@ -0,0 +1,1180 @@
+	oldAtt, att := oldObj.(*netv1a1.NetworkAttachment), obj.(*netv1a1.NetworkAttachment)
+	klog.V(5).Infof("Local NetworkAttachments cache: notified of update from %#+v to %#+v", oldAtt, att)
+
+	// Enqueue if the UID changed because if a local NetworkAttachment is
+	// deleted and replaced the status.hostIP field of the newer attachment is
+	// set to "", and the connection agent has to write back the correct value.
+	// Also, the only fields affecting local network interfaces handling that
+	// can be seen changing by this function are status.ipv4 and
+	// status.addressVNI, so enqueue if they changed.
+	if oldAtt.UID != att.UID || oldAtt.Status.IPv4 != att.Status.IPv4 || oldAtt.Status.AddressVNI != att.Status.AddressVNI {
+		ca.queue.Add(parse.AttNSN(att))
+	}
+}
+
+func (ca *ConnectionAgent) onLocalAttDelete(obj interface{}) {
+	att := parse.Peel(obj).(*netv1a1.NetworkAttachment)
+	klog.V(5).Infof("Local NetworkAttachments cache: notified of removal of %#+v", att)
+
+	attNSN := parse.AttNSN(att)
+	ca.updateL1VNStateForLocalAtt(attNSN, false)
+	ca.queue.Add(attNSN)
+}
+
+func (ca *ConnectionAgent) updateL1VNStateForLocalAtt(att k8stypes.NamespacedName, attExists bool) {
+	ca.l1VirtNetsState.Lock()
+	defer ca.l1VirtNetsState.Unlock()
+
+	attSeenInCaches := ca.l1VirtNetsState.attToSeenInCaches[att]
+
+	if attExists {
+		if attSeenInCaches == nil {
+			attSeenInCaches = make(map[cacheID]struct{}, 1)
+			ca.l1VirtNetsState.attToSeenInCaches[att] = attSeenInCaches
+		}
+		attSeenInCaches[localAttsCacheID] = struct{}{}
+		return
+	}
+
+	delete(attSeenInCaches, localAttsCacheID)
+	if len(attSeenInCaches) == 0 {
+		delete(ca.l1VirtNetsState.attToSeenInCaches, att)
+	}
+}
+
+func (ca *ConnectionAgent) syncPreExistingNetworkInterfaces() error {
+	// Start all the remote attachments informers because to choose whether to
+	// keep a pre-existing remote network interface we need to look for a
+	// remote network attachment that can own it in the informer cache for the
+	// VNI of the network interface.
+	err := ca.startRemoteAttsInformers()
+	if err != nil {
+		return fmt.Errorf("failed to sync pre-existing network interfaces: %s", err.Error())
+	}
+
+	ifcs, err := ca.listPreExistingNetworkInterfaces()
+	if err != nil {
+		return fmt.Errorf("failed to sync pre-existing network interfaces: %s", err.Error())
+	}
+
+	for _, ifc := range ifcs {
+		err = ca.syncPreExistingNetworkInterface(ifc)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (ca *ConnectionAgent) startRemoteAttsInformers() error {
+	localAtts, err := ca.localAttsLister.List(k8slabels.Everything())
+	if err != nil {
+		return fmt.Errorf("failed to list local network attachments: %s", err.Error())
+	}
+
+	for _, att := range localAtts {
+		// Adding the local NetworkAttachment to the L2 virtual network state
+		// entails starting the remote attachments informer for the attachment's
+		// vni.
+		ca.addLocalAttToL2VNState(parse.AttNSN(att), att.Status.AddressVNI)
```

Comment:
A few clarifications:
1. my previous comment state that you are advocating for checking the error. You are not, sorry, I misread. You are advocating for explaining why the error is not checked in a comment. 40f379e adds such comments.

2. I brought up consistency of cache-based list results with etcd content assuming that there will never be two NetworkAttachments with an IP address, same VNI and different namespaces in etcd; if that is true and consistency is guaranteed that error will never happen in the code above. But I thought about a scenario where two NetworkAttachments which have already been assigned an IP address exist in etcd at the same time with same VNI and different namespace (during a transient), hence regardless of consistency of cache-based list results the error can happen. I believe it's ok not to check it because the code above does not know which namespace is the most recent one for the VNI being considered (or, it cannot know with a simple algorithm, or, if it knows it cannot undo its previous actions consequently with a simple algorithm), this favors simplicity over accuracy and is consistent with other choices made during the initial sync, such as the one of keeping the interface that was matched first in case the owner for a pre-existing interface was already matched to one. If the wrong choice is taken, when the workers start the mistake is rectified. The added comment summarizes these arguments.

Reply:
> IIRC, the reason for a live list in the subnet validator has to do with freshness (really, forcing ordering among operations) rather than coherence within a list result.

Assume two conflicting subnets S1 and S2 are created concurrently. etcd serializes all writes therefore subnet ops are fully ordered: assume S1 is created first. Regardless of whether live or cache-based lists are used, the worker validating S1 might not see S2 because S2 is created after S1. Now consider the worker validating S2. If it uses live lists, the list hits etcd after S2 was created, which is after S1 was created: list results either include S1 or S1 has been deleted in the meantime. in a nutshell:

`P1: it's impossible to miss S1 if it exists.`

If cache-based list results are consistent with etcd, P1 holds for cache-based lists as well, because the etcd state that populated the cache when the cache-based list was done already comprised S2, so it comprised S1 as well (assuming it was not deleted in the meantime). Hence the subnet validator CAN use cache-based lists. If cache-based list results are not consistent with etcd, P1 does not hold and live lists must be used. This is assuming pagination of LISTs is not done.

Reply:
Since informer's caches are not consistent with the authoritative state of the cluster, the subnet validator must use live lists to ensure correctness.

Reply:
This discussion went a bit off the rails at https://github.com/MikeSpreitzer/kube-examples/pull/81#discussion_r323662370 .  The qualification "once HasSynced() returns true, anyway" was unnecessary caution on my part.  Every query to an informer's cache is answered entirely from one collection state, regardless of when the query is made.  I hope the comments on informers make that clear now.  Our problem here is that the collection of NetworkAttachment objects is not constrained to consistently relate VNI to namespace.

Reply:
> Our problem here is that the collection of NetworkAttachment objects is not constrained to consistently relate VNI to namespace.

Yes, let's no longer talk about that in this thread since comments further down this PR discuss that.

> The qualification "once HasSynced() returns true, anyway" was unnecessary caution on my part. Every query to an informer's cache is answered entirely from one collection state, regardless of when the query is made. I hope the comments on informers make that clear now. 

I'm confused by the conversation that led to the informers comments rather than the informers comments themselves. What confuses me is that you and Daniel Smith seem to have two different notions of collection state, and such divergence is never resolved (it's very likely that I'm misunderstanding something here). Daniel Smith states that [an informer's collection state is, or was at some point in the past, the authoritative state of the cluster as returned by the API server](https://github.com/kubernetes/kubernetes/pull/77636#discussion_r282549940) (i.e. stored on etcd). That is false, and you correctly [pointed that out](https://github.com/kubernetes/kubernetes/pull/77636#discussion_r282588751). He [agrees not to explicitly mention that property in the informers' comments, but re-states that it is true](https://github.com/kubernetes/kubernetes/pull/77636#discussion_r282595214). Then [your last comment](https://github.com/kubernetes/kubernetes/pull/77636#discussion_r282598321) seems to suggest that you also believe that property holds true. I'm trying to understand whether someone believes that informer's lists return a (subset of a) collection of objects with their states that was actually stored on etcd at some point in time; that is false and assuming it's true is a critical mistake.

Reply:
Agreed.  There was confusion in the past.  I think now we are clear that, as seen by a given query, an informer's local cache does not necessarily reflect a collection state that existed in etcd.  I think the current KOS code does not make a bogus assumption in this regard.

Reply:
Agreed.